New action type - D3 security SOAR

d3-naiji opened the pull request:
## Summary

Introduces a new built-in action type and sends the ElasticSearch alerts to D3 Security SOAR for incident response, workflow automation and deep investigation. This integrated alert action allows you to create Events or trigger Playbook workflow actions automatically in the D3 Security SOAR platform. 

The new action type and the action connector with D3 Security SOAR are as follows: 

![D3Security connector card](https://i.imgur.com/pbmXBVy.png)

![D3Security connector form](https://i.imgur.com/bjluUsl.png)

and action params form available in Create Alert form:
![D3Security action form](https://i.imgur.com/TwioSA0.png)

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)



### For maintainers

v8.3.3
Release_note:feature

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

cnasikas commented:
Hi @d3-naiji! Thank you so much for your contribution. I want to ask you if you want to support the connector in Cases? I am asking because Cases use a different flow that alerts. You can create an incident, update the incident, add comments, etc. To support both cases, the code needs to also consider the case user flow.

pmuellr commented:
Thanks for the PR, @d3-naiji!  A couple things we'll need, beyond questions that Christos mentioned ^^^, if the desire is to ship this as-is vs as a case connector.

- the CLA must be signed; do that here: https://www.elastic.co/contributor-agreement - it's the same link from the build comment ^^^
- we'll need some "function tests"

The last connector we added was xMatters in PR https://github.com/elastic/kibana/pull/122357 ; the "function test" I mentioned above are in the `x-pack/test/alerting_api_integration` directory, and we'd want the same kind of tests that were added for xMatters, so hopefully you can follow that PR to figure out what to do.  Don't hesitate to ask us more questions!

d3-naiji commented:
> [cnasikas](/cnasikas)
Hi cnasikas, thanks for the quick review. I think we can remove the "Case" tag for now. And we do plan to support cases in the future.

d3-naiji commented:
> Thanks for the PR, @d3-naiji! A couple things we'll need, beyond questions that Christos mentioned ^^^, if the desire is to ship this as-is vs as a case connector.
> 
> * the CLA must be signed; do that here: https://www.elastic.co/contributor-agreement - it's the same link from the build comment ^^^
> * we'll need some "function tests"
> 
> The last connector we added was xMatters in PR #122357 ; the "function test" I mentioned above are in the `x-pack/test/alerting_api_integration` directory, and we'd want the same kind of tests that were added for xMatters, so hopefully you can follow that PR to figure out what to do. Don't hesitate to ask us more questions!

Hi @pmuellr , thanks for the information. We have signed CLA already. And will add the function test in the next commit.
An update for Christos's question: We will not make a case connector for now and will remove the "Case" tag. Will ping you when it's complete.

cnasikas commented:
> > Thanks for the PR, @d3-naiji! A couple things we'll need, beyond questions that Christos mentioned ^^^, if the desire is to ship this as-is vs as a case connector.
> > 
> > * the CLA must be signed; do that here: https://www.elastic.co/contributor-agreement - it's the same link from the build comment ^^^
> > * we'll need some "function tests"
> > 
> > The last connector we added was xMatters in PR #122357 ; the "function test" I mentioned above are in the `x-pack/test/alerting_api_integration` directory, and we'd want the same kind of tests that were added for xMatters, so hopefully you can follow that PR to figure out what to do. Don't hesitate to ask us more questions!
> 
> Hi @pmuellr , thanks for the information. We have signed CLA already. And will add the function test in the next commit. An update for Christos's question: We will not make a case connector for now and will remove the "Case" tag. Will ping you when it's complete.

Thank you @d3-naiji. I would like to let you know that we have created a framework to aid developers in creating connectors in case you are interested. You can find the documentation here https://github.com/elastic/kibana/tree/main/x-pack/plugins/actions/server/sub_action_framework. I would suggest using it if you want to be "case" compatible in the future.

d3-naiji commented:
Hi @pmuellr, I have added the "funtion test" in the x-pack/test/alerting_api_integration directory and followed the instructions from https://www.elastic.co/guide/en/kibana/master/development-tests.html#development-functional-tests
So, I tried to run command's bellow
node scripts/functional_tests_server, which was successfully. But, when I tried to run node scripts/functional_test_runner, it required a --config, please see the image bellow.
![image](https://user-images.githubusercontent.com/109116834/182478059-be2e439d-ac3b-43cb-acc8-4899455cd24e.png)
After I added the config path, it says Error: attempted to use the "es" service to fetch Elasticsearch version info but the request failed: ConnectionError: write EPROTO 4771253760:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:
![image](https://user-images.githubusercontent.com/109116834/182478976-b3f2888a-fc14-4b1f-87ec-09f228577d55.png)
I have searched up, but still couldn't fix the problem.
Do you know how can I test the "funtion test" files or if you have any references or tips for that?
I look forward to hearing from you, thank you!

EricDavisX commented:
Hi, @d3-naiji I wanted to reach out, is there a point of contact we can talk to about setting up a real 'test instance' of the D3 Soar code base for on-going end to end validation?  I can't commit to how fast we can work it, but I'm a good resource to kick that chat off.  I can read up on the docs, too.  TIA!

d3-naiji commented:
> Hi, @d3-naiji I wanted to reach out, is there a point of contact we can talk to about setting up a real 'test instance' of the D3 Soar code base for on-going end to end validation? I can't commit to how fast we can work it, but I'm a good resource to kick that chat off. I can read up on the docs, too. TIA!

Hi @EricDavisX , thanks for the update. I will prepare a document and a demo video about how to use the Kibane Connector to create events or trigger playbook flows in D3 SOAR. I will send you the documents and the 'test instance' in an email this week.

d3-naiji commented:
> Hi, @d3-naiji I wanted to reach out, is there a point of contact we can talk to about setting up a real 'test instance' of the D3 Soar code base for on-going end to end validation? I can't commit to how fast we can work it, but I'm a good resource to kick that chat off. I can read up on the docs, too. TIA!

@EricDavisX The document and D3 SOAR "test instance" info was sent to your email.

EricDavisX commented:
Confirming as I understand it that @MindyRS 's team will take this on from testing side.  I have forwarded the note to her I received.  @MadameSheema and @asnehalb CC.  Regards!

stephmilovic commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

stephmilovic commented:
@elasticmachine merge upstream

stephmilovic commented:
@d3-naiji could you please update the PR and resolve the conflicts? I will be reviewing this PR today/this week. Thank you

semd commented:
buildkite test this

semd commented:
buildkite test this

semd commented:
Hi @d3-naiji,
Sorry, we have been busy during the last couple of weeks due to the release schedule. 
There have been new connectors merged, which has caused some more conflicts to appear, they should be easy to fix.

semd commented on x-pack/plugins/actions/server/builtin_action_types/d3security.ts:
```diff
@@ -0,0 +1,295 @@
+    }),
+    supportedFeatureIds: [
+      AlertingConnectorFeatureId,
+      // CasesConnectorFeatureId,
```
please remove comment

semd commented on x-pack/plugins/triggers_actions_ui/README.md:
```diff
@@ -1,4 +1,4 @@
-# Kibana Alerts and Actions UI
+# Kibana Alerts an3 Actions UI
```
?

semd commented on x-pack/test/alerting_api_integration/common/fixtures/plugins/actions_simulators/server/d3security_simulation.ts:
```diff
@@ -0,0 +1,127 @@
+            return validateAuthentication(credentials, response);
+          case 'success_post_method':
+            return validateRequestUsesMethod(request.method ?? '', 'post', response);
+          // case 'success_put_method':
+          //   return validateRequestUsesMethod(request.method ?? '', 'put', response);
```
Is PUT a valid method?

semd reviewed (commented):
New code need to be formatted using `prettier`

semd commented:
@EricDavisX @d3-naiji 
Would it be possible to have the test instance info sent to my Elastic address (`sergi.massaneda@elastic.co`) as well? I will need it in order to approve the PR.

d3-naiji commented:
> @EricDavisX @d3-naiji Would it be possible to have the test instance info sent to my Elastic address (`sergi.massaneda@elastic.co`) as well? I will need it in order to approve the PR.

Hi @semd , have you received the test instance? Do you want me to send it again? 
And is there any update for the review? We submitted another pull request #144955 for merging the conflicts of the file structure change. Please let me know if you need another other information from us.

d3-naiji commented:
Hi @EricDavisX @semd,
In case it was lost, I have re-sent the test instance information to your email. Please let me know if you have any trouble accessing it or if there are any further questions or comments that you may have.

d3-bensong commented:
@elasticmachine merge upstream
